File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/factory/LoginViewModelFactory.java
package com.platdmit.simplecloudmanager.app.vm.factory;

import com.platdmit.simplecloudmanager.domain.helpers.ActualApiKeyService;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class LoginViewModelFactory<Repo> extends ViewModelProvider.NewInstanceFactory {
    private Repo mRepo;
    private Class mRepoInterface;
    private ActualApiKeyService mActualApiKeyService;

    public LoginViewModelFactory(Repo repo, Class<Repo> repoInterface, ActualApiKeyService actualApiKeyService) {
        super();
        mRepo = repo;
        mRepoInterface = repoInterface;
        mActualApiKeyService = actualApiKeyService;
    }

    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        try {
            return (T)modelClass.getConstructor(mRepoInterface, ActualApiKeyService.class).newInstance(mRepo, mActualApiKeyService);
        } catch (Exception e){
            try {
                return modelClass.newInstance();
            } catch (InstantiationException | IllegalAccessException g) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, g);
            }
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/factory/LoginViewModelFactory.kt
package com.platdmit.simplecloudmanager.app.vm.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import com.platdmit.simplecloudmanager.domain.helpers.ActualApiKeyService

class LoginViewModelFactory<Repo>(private val repo: Repo, private val repoInterface: Class<Repo>, private val actualApiKeyService: ActualApiKeyService) : NewInstanceFactory() {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return try {
            modelClass.getConstructor(repoInterface, ActualApiKeyService::class.java).newInstance(repo, actualApiKeyService)
        } catch (e: Exception) {
            try {
                modelClass.newInstance()
            } catch (g: InstantiationException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            } catch (g: IllegalAccessException) {
                throw RuntimeException("Cannot create an instance of $modelClass", g)
            }
        }
    }
}